docs/proposals/rebalancing-clusters-across-shards-dynamically.md

```diff
@@ -0,0 +1,144 @@
+
+The heartbeat interval will be a configurable parameter initialized while setting up the application controller. This way, users will be able to control the frequency at which they want the heartbeat process to take place.
+
+Anytime the number of healthy replicas of application controllers is different from the number of application controllers to shard mappings, we would re-distribute the clusters among the healthy replicas again. We can summarize the above statement using the below formula:
+
+```
+Number of Replicas ≠ Count of {Application Controller, Shard} mapping
+```
+
+The below logic can be used to perform application controller to shard assignment:
+
+1) If a new application controller is added, that is, a new shard is added, we would perform the re-distribution of clusters among the shards with the existing sharding algorithm being used.
+
+2) In scenarios when one of the application controllers is identified to be unhealthy, the rest of the application controllers will re-distribute the load(clusters) managed by the application controller and start managing the orphaned applications. 
+
+How will this work? 
+* The application controller will query Redis for the status of all the application controllers and last updated heartbeat timestamp.
+* It will check if any application controller is flagged as Unhealthy or has not updated its status in Redis during the heartbeat process for a certain period of time.
+* If the status for an application controller was already flagged as Unhealthy, it will pick one of the clusters from its list and assign the cluster to itself.
+* If the status is not flagged and an application controller has not updated the last active timestamp in a long time, then we mark the Application Controller as Unhealthy. 
+
+This will continue to happen till the time the list of clusters is marked as empty. As soon as the list is marked as empty, the entry of the shard/application-controller is removed from Redis.
+
+*Note:* We will continue to use the cluster to shard assignment approach being used today.
+
+### Pros
+* Every Application Controller would be able to take action on finding the distribution of load.
+* Every Application Controller will monitor the status of Unhealthy shard and would be able to take action or flag for action.
+
+### Cons
+
+* Possibility of race conditions while flagging the shard as Unhealthy during the heartbeat process. Although this can be handled using the [distributed locks](https://redis.io/docs/manual/patterns/distributed-locks/) in Redis.
+
+* In scenarios when Redis becomes unavailable, the heartbeat mechanism will pause working till the redis comes back online again. This will also pause the dynamic redistribution of clusters till Redis comes back online. The redistribution of clusters will be triggered again when Redis comes back online.
```

Comment:
Maybe we could rely less in Redis if we had leader election in place. In this case we could instead, maintain an InMemory cache of the sharding state. This would be synced in Redis mainly so it can be picked by the new leader in case of current leader crashes.

Reply:
How about we use configmaps for this instead of redis, k8s would prevent concurrent updates to the configmaps so we can rely on that for handling race conditions while we also remove the issue of relying on redis too much.